### DISCUSSION TITLE
Fully utilize autocompletion ordering for hrsh7th/nvim-cmp

### DISCUSSION THREAD

**Discussion by rochala** [scalameta/nvim-metals#539] [Show and tell]
Hello everyone!
As you know, `nvim-metals` uses Language Server Protocol to provide IDE capabilities.
The one I'm going to create discussion about is `completion` with specifications available [here](
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion).

`nvim-metals` plugin is not responsible for displaying the tooltips, only fetching them from the server. This is where things get interesting. LSP provides a way to return completion order by setting the `sortText` string value. If it is undefined, all ordering and sorting is handled by the client, in this case by `hrsh7th/nvim-cmp` plugin.

The default sorting configuration matches the completions in the following way:
```lua
    sorting = {
      priority_weight = 2,
      comparators = {
        compare.offset,
        compare.exact,
        compare.score,
        compare.recently_used,
        compare.kind,
        compare.sort_text,
        compare.length,
        compare.order,
      },
    },
```

As you see, the `sort_text` is present in this configuration and by default it should work as intended... but it isn't.

As an example, we will check the completion is `build.sbt` file:
<img width="705" alt="image" src="https://user-images.githubusercontent.com/48657087/221614353-0e0481c8-9087-4c98-a43e-496ae5dc67d9.png">
And here the `package` is a preselected value. If we check the ordering returned from metals we can see that it returns `package` with `sortText` = `"00000"` so it should be our first result.

```
[Trace - 04:59:28 PM] Sending response 'textDocument/completion - (4)'. Processing request took 400ms
Result: {
  "isIncomplete": true,
  "items": [
    {
      "label": "`package`: TaskKey[File]",
      "kind": 12,
      "detail": ": TaskKey[File]",
      "preselect": true,
      "sortText": "00000",
      "filterText": "package",
      "insertTextFormat": 2,
      "textEdit": {
        "range": {
          "start": {
            "line": 10,
            "character": 0
          },
          "end": {
            "line": 10,
            "character": 1
          }
        },
        "newText": "`package`"
      },
      "data": {
        "symbol": "sbt/Keys.package.",
        "target": "file:/Users/jrochala/Projects/playground/project/?id\u003dplayground-build",
        "kind": 0
      }
    },
```

So let's get a quick summary before going further:
- completions are not ordered properly,
- this leads to a preselected item to being in the middle of the completion list, which is super annoying

Let's fix this, so we can have the correct ordering from the server:

<img width="375" alt="image" src="https://user-images.githubusercontent.com/48657087/221615560-e39566f6-d000-416c-a3af-6e6d9b5b53b7.png">

For this, we can't easily order only by `sortText` as metals optimizes the calls to the server.
In short, to not recompute file on every letter typed, the response from completion also contains the `isIncomplete` field. It only does recompute when it is true, and usually 3 letters are enough to get all available completions and further letters typed no longer are sorted by server...
This is what happens after the 4th letter:

https://user-images.githubusercontent.com/48657087/221618235-c02ce97a-6b93-4d42-9a97-86ad2a9936ec.mov

As you can see, our selection randomly jumped to the middle of the list. This is caused because our ordering has higher precedence of `compare.offset` than our `compare.sort_text` and it doesn't refresh preselect!

The fix is to add the following lines to your `nvim-cmp` configuration (it will affect other LSP configurations, but I leave it to you to figure out how to enable it only for metals):
```  
cmp.setup({
  preselect = cmp.PreselectMode.None, -- disable preselection
  sorting = {
    priority_weight = 2,
    comparators = {
      compare.offset, -- we still want offset to be higher to order after 3rd letter
      compare.score, -- same as above
      compare.sort_text, -- add higher precedence for sort_text, it must be above `kind`
      compare.recently_used,
      compare.kind,
      compare.length,
      compare.order,
    },
  },
  -- if you want to add preselection you have to set completeopt to new values
  completion = {
    -- completeopt = 'menu,menuone,noselect', <---- this is default value,
    completeopt = 'menu,menuone', -- remove noselect
  },
})
```
In my opinion, this is not a bug in `hrsh7th/nvim-cmp` but rather a problem with the unique metals client implementation. This should improve your completions.

**Comment by ckipp01** [comment]
Thanks a lot for this @rochala!
